### ISSUE TITLE
RequestTimeout: timed out due to a ping-pong keepalive missing on time

### ISSUE THREAD

**Issue by nileio** [ccxt/ccxt#15343]
- OS: linux
- Programming Language version: node v16.17.1
- CCXT version: 2.0.31

```
//`watchTrades` on Binance every second minute the error is produced and simply things stop suddenly and can't figure out how to even handle the error gracefully or re-subscribe
client = new ccxt.pro['binance']({
      newUpdates: true
    });
while (true) {
              try {
                const trades = await client.watchTrades(symbol);
                //doing things here
              } catch (e) {
                // debugger;
                console.log(e);
                // do nothing and retry on next loop iteration
                // throw e // uncomment to break all loops in case of an error in any one of them
                // break // you can also break just this one loop if it fails
              }
}
```

```
RequestTimeout: Connection to wss://stream.binance.com:9443/ws timed out due to a ping-pong keepalive missing on time
```

**Comment by ttodua** [comment]
Hi @nileio , can you please tell if you still experience that issue?  maybe it was some temporary network-related or etc, as I've tested (running already half an hour) and it doesn break even once.

**Comment by wunderb1t** [comment]
Hi @ttodua, we had also experienced that issue (from Oct 25, 2022 at 13:27:32.333 to Oct 25, 2022 at 20:40:15.209) but on Bybit and on `watchTickers` method.
Message:
`Connection to wss://stream.bybit.com/realtime_public timed out due to a ping-pong keepalive missing on time`

Also we've noticed that every time after ~10 hours of socket connection was up this issue appears again and only thing that helps to resolve it is restart the node instance.
And we use per each exchange one ccxt instance for sockets without recreating them after socket connection lost maybe that is the case too.

**Comment by nileio** [comment]
@ttodua it is intermittent.. It happens when subscribing to multiple markets which is allowed by Binance for example, we should be able to subscribe to up to 1024 markets with one connection. Now I am trying to dig in the code to understand the ping-pong logic. Binance actually abides by WS standards which is great, it only requires `pong` when `ping` is recieved, and ws should emit both `ping` and `pong`.. https://datatracker.ietf.org/doc/html/draft-ietf-hybi-thewebsocketprotocol-07#section-4.5.3

**Comment by nileio** [comment]
> Hi @nileio , can you please tell if you still experience that issue? maybe it was some temporary network-related or etc, as I've tested (running already half an hour) and it doesn break even once.

Please test it after subscribing for multiple markets, that is , 500 markets or so

**Comment by caiusCitiriga** [comment]
I am experiencing this issue too. Trying to subscribe to 338 markets trades and intermittently  receive this error. 

Also, sometimes appears this error too:
`connection closed by remote server, closing code 1008`;

This can be related to: https://github.com/ccxt/ccxt/issues/12031#issuecomment-1312796717 ?

As @nileio said, Binance should let you subscribe 1024 streams from a single connection. Any idea?

**Comment by caiusCitiriga** [comment]
@ttodua I can confirm that with 100 markets subscriptions from a single connection doesn't throw errors. With that being said, this is kinda critical. I'm aware that Binance is the most generous exchange in terms of data provisioning, but their documentation states that:

![Screenshot 2022-11-14 at 10 24 59](https://user-images.githubusercontent.com/12595721/201623613-e22a2a89-fd8c-4abd-ae44-fbaec134b836.png)

So, am I missing something? The ccxt manual says that the websocket connection is reused if already existing. So, since the `watchTrades` runs in a loop, I expect that ccxt adds a new subscription for a given symbol whenever a new symbol (that is not already subscribed) is passed to the function.

So, why does it throw errors if the number of markets to watch is way lower that the 1024 that binance says you can use? 

I've also tried to add a delay at the top of the `loop()` to throttle the subscriptions, but after a while or in some cases after a minute it blows up anyway.

I've also found this issue that seems related: 
https://github.com/ccxt/ccxt/issues/9907

where @kroitor says that opening more than 50/100 subscriptions is not safe. But why? 
I'm really confused

**Comment by ttodua** [comment]
@carlosmiei @pcriadoperez @sc0Vu do you have any thoughts on this?

**Comment by pcriadoperez** [comment]
Hi,

So I think there are two different issues here, the 1008 "Too many requests" error and the ping-pong keep alive error.

- For the 1008 error, after a bunch of testing I found that the issue is we are only using one stream for all of our requests to binance. So even though they allow 1024 streams, we currently only use one, so when receiving too many messages over the stream it throws the 1008 error. I want to dig in more to this but I will be posting a PR with a solution shortly. This will also solve issue #12031 

- Second we have the ping-pong error. Unfortunately I still have not been able to reproduce this error.
@nileio , I'm not sure this answers your comment above, but we do change the ping/pong logic depending on the exchange if needed. My suggestion here is, I'll implement the multiple streams on binance which will remove the stress on each client, and as you mention it only happens when subscribing to a lot of markets, hopefully this will also solve the ping-pong issue. If not happy to continue looking into it.

**Comment by Kinzowa** [comment]
Hi, I just got the ping-pong issue with Bybit while listening to 2 markets `watch_tickers` and 2 private methods simultanously: `watchMyTrades` and `watchOrders`. From my experience with CCXT the websocket wrappers throw error with all exchanges I've tried.

`timed out due to a ping-pong keepalive missing on time`

**Comment by caiusCitiriga** [comment]
> Hi,
> 
> So I think there are two different issues here, the 1008 "Too many requests" error and the ping-pong keep alive error.
> 
> * For the 1008 error, after a bunch of testing I found that the issue is we are only using one stream for all of our requests to binance. So even though they allow 1024 streams, we currently only use one, so when receiving too many messages over the stream it throws the 1008 error. I want to dig in more to this but I will be posting a PR with a solution shortly. This will also solve issue [Connection closed by remote server, closing code 1008 - compare with cryptofeed library #12031](https://github.com/ccxt/ccxt/issues/12031)
> * Second we have the ping-pong error. Unfortunately I still have not been able to reproduce this error.
>   @nileio , I'm not sure this answers your comment above, but we do change the ping/pong logic depending on the exchange if needed. My suggestion here is, I'll implement the multiple streams on binance which will remove the stress on each client, and as you mention it only happens when subscribing to a lot of markets, hopefully this will also solve the ping-pong issue. If not happy to continue looking into it.

@pcriadoperez Ok, knowing that ccxt uses only one stream for binance, the 1008 error now makes sense. So I'll wait for the PR and then I'll try again. Can you ping me here once it's merged? Thank you!

**Comment by caiusCitiriga** [comment]
hi @pcriadoperez I've updated the ccxt library to the last version. Indeed the 1008 error seems to have been solved, but unfortunately the "timed out due to a ping-pong keepalive missing on time" error is still present. 

any ideas? 😞

**Comment by caiusCitiriga** [comment]
### Update:

It happens even with only 1 market subscription, while building 12 timeframes. Although I'm not totally sure about this, but it seems somehow related to the amount of workload of the process, but why? How could be this related? And why would ccxt miss a ping-pong in time due to this?

**Comment by caiusCitiriga** [comment]
wow, awesome support. As always

**Comment by frosty00** [comment]
> 
Update:
It happens even with only 1 market subscription, while building 12 timeframes. Although I'm not totally sure about this, but it seems somehow related to the amount of workload of the process, but why? How could be this related? And why would ccxt miss a ping-pong in time due to this?

Yeah the pongs are sent in response to a ping and they are scheduled through the event loop. so any code that does not release the event loop would cause this to happen. 

> wow, awesome support. As always

We are working on other issues, and fixing bugs elsewhere in the lib. Sorry that your bug was not responded to.

**Comment by nileio** [comment]
@caiusCitiriga I have a feeling something fundamental is going on there .. throttling subscriptions to Binance might simply be the solution (managed by the backend).. I'd stay away for now would find time to get back to my project over xmas

**Comment by ttodua** [comment]
@nileio I am not working on this subject, you might tag @carlosmiei / @pcriadoperez / @sc0Vu to find out who is tracking this issue.

**Comment by DOABrownie** [comment]
I'm also getting this error when watching only one order books on Kucoin. Ping pong missing error gets through after approx 1-3min of the order book polling in console.

**Comment by Kinzowa** [comment]
> Hi,
> 
> So I think there are two different issues here, the 1008 "Too many requests" error and the ping-pong keep alive error.
> 
> * For the 1008 error, after a bunch of testing I found that the issue is we are only using one stream for all of our requests to binance. So even though they allow 1024 streams, we currently only use one, so when receiving too many messages over the stream it throws the 1008 error. I want to dig in more to this but I will be posting a PR with a solution shortly. This will also solve issue [Connection closed by remote server, closing code 1008 - compare with cryptofeed library #12031](https://github.com/ccxt/ccxt/issues/12031)
> * Second we have the ping-pong error. Unfortunately I still have not been able to reproduce this error.
>   @nileio , I'm not sure this answers your comment above, but we do change the ping/pong logic depending on the exchange if needed. My suggestion here is, I'll implement the multiple streams on binance which will remove the stress on each client, and as you mention it only happens when subscribing to a lot of markets, hopefully this will also solve the ping-pong issue. If not happy to continue looking into it.

I monitor the markets of 18 exchanges and am getting the ping-pong issue after few minutes with all exchanges except Binance. Could the patch you applied for Binance explain the absence of ping pong errors on this exchange?

@carlosmiei / @pcriadoperez / @sc0Vu

**Comment by caiusCitiriga** [comment]
@yyqiekenao yeah, stop using ccxt

**Comment by karambaq** [comment]
Same problem. Someone find the solution?

**Comment by nileio** [comment]
Nop - I moved on already !!

**Comment by ragamufin** [comment]
In my case this was happening with `watchBalance()`. It would happen sporadically giving the error `timed out due to a ping-pong keepalive missing on time` stopping my code from continuing. Sometimes I would get the error `NetworkError connection closed by remote server, closing code 1006`

As explained in https://github.com/ccxt/ccxt/issues/7951#issuecomment-724762855 and https://github.com/ccxt/ccxt/issues/17162#issuecomment-1468196614, putting a `try` `catch` around it I am able to catch the error and then recover from it similar to the following:

```javascript
while(true) {
	console.log('*** watchBalance ***')
	try {
		const balances = await exchange.watchBalance()
		console.log('balances:', balances)
	} catch(err) {
		console.warn(err)
	}
}
```

Since I'm running a loop, if there's an error, it gets caught and then in the next iteration of the loop the connection is restored. You could put any recovery/cleanup code you need in the `catch` block.  This may help anyone having these errors with any of the methods that run through websockets such as in the following issues:

https://github.com/ccxt/ccxt/issues/15338
https://github.com/ccxt/ccxt/issues/14557
https://github.com/ccxt/ccxt/issues/14086
https://github.com/ccxt/ccxt/issues/22493
https://github.com/ccxt/ccxt/issues/7858
https://github.com/ccxt/ccxt/issues/10955

**Comment by bznsix** [comment]
> 在我的例子中，这种情况发生在 `watchBalance()` 时，偶尔会出现 `timed out due to a ping-pong keepalive missing on time` 错误，导致代码无法继续。有时我还会收到 `NetworkError connection closed by remote server, closing code 1006` 或 `failed due to a connection timeout` 的错误信息。
> 
> 正如在 [#7951 (comment)](https://github.com/ccxt/ccxt/issues/7951#issuecomment-724762855) 和 [#17162 (comment)](https://github.com/ccxt/ccxt/issues/17162#issuecomment-1468196614) 中解释的那样，在它周围加上 `try` `catch` 就能捕捉到错误，然后类似于下面这样恢复：
> 
> ```js
> while(true) {
> 	console.log('*** watchBalance ***')
> 	try {
> 		const balances = await exchange.watchBalance()
> 		console.log('balances:', balances)
> 	} catch(err) {
> 		console.warn(err)
> 	}
> }
> ```
> 
> 由于我运行的是一个循环，如果出现错误，就会被捕获，然后在循环的下一次迭代中恢复连接。你可以在 `catch` 代码块中加入任何需要的恢复/清理代码。这可能会对通过 webockets 运行的任何方法（如以下问题）出现错误的人有所帮助：
> 
> #15338 #14557 #14086 #22493 #7858 #10955

Using try catch,will ccxt auto rebuild the connection?

**Comment by ragamufin** [comment]
Yes, if it's in a loop as in the example. In the `catch` block you can do whatever you need to do in order to handle other things that might break depending on your code. Since add the `try` - `catch` I've seen the errors come up and immediately the connection is re-established

**Comment by bznsix** [comment]
> Yes, if it's in a loop as in the example. In the `catch` block you can do whatever you need to do in order to handle other things that might break depending on your code. Since add the `try` - `catch` I've seen the errors come up and immediately the connection is re-established

Yes,trying to fix my code use try catch,and it will auto reconnect.

**Comment by dhamoblaze** [comment]
> This error persists. I’m handling it with a try-catch block, but I keep encountering "timed out due to a ping-pong keepalive missing on time" errors.
> 
> I get this error every 10 seconds, occurring for both Deribit and Bybit exchanges. All my WebSocket connections throw this error; I need to reconnect them each time.
> 
> I’ve split my WebSocket connections into batches, each handling 50 symbols, but the issue still occurs.
> 
> exchanged = "bybit"
> exchanges[exchangeId] = new ccxtpro[exchangeId]({
>     enableRateLimit: true,
>     options: {
>         defaultType: 'option',
>         adjustForTimeDifference: true,
>         keepAlive: true,
>         pingInterval: 5000,
>         pingTimeout: 20000,
>     },
> });

Have you found a solution?

**Comment by caiusCitiriga** [comment]
@dhamoblaze if you consider switching to another solution, https://www.npmjs.com/package/binance is the best in class. And from the same author you get all these packages:

<img width="519" height="530" alt="Image" src="https://github.com/user-attachments/assets/1e1cddf4-152d-4c02-9de9-0c9326f43aab" />

**Comment by dhamoblaze** [comment]
> [@dhamoblaze](https://github.com/dhamoblaze) if you consider switching to another solution, https://www.npmjs.com/package/binance is the best in class. And from the same author you get all these packages:
> 
> <img alt="Image" width="519" height="530" src="https://private-user-images.githubusercontent.com/12595721/487343138-1e1cddf4-152d-4c02-9de9-0c9326f43aab.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc1OTcyNjQsIm5iZiI6MTc1NzU5Njk2NCwicGF0aCI6Ii8xMjU5NTcyMS80ODczNDMxMzgtMWUxY2RkZjQtMTUyZC00YzAyLTlkZTktMGM5MzI2ZjQzYWFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTExVDEzMjI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ4NjVlYzBlYjlmMmQzMDRkMjQ0ZjA2YmQ5NzMyMWZlMzMzNzgzMTMyNzc4MDBjZTYzNDhlY2EyODk3Nzk2NmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._3ap5_HZijArMw-Xvvf-FpfTbCbtLlkg0IJxUTqUOKY">

We have connected to the top 20 exchanges in our product, so CCXT is the better option

**Comment by cenaav** [comment]
> > This error persists. I’m handling it with a try-catch block, but I keep encountering "timed out due to a ping-pong keepalive missing on time" errors.
> > I get this error every 10 seconds, occurring for both Deribit and Bybit exchanges. All my WebSocket connections throw this error; I need to reconnect them each time.
> > I’ve split my WebSocket connections into batches, each handling 50 symbols, but the issue still occurs.
> > exchanged = "bybit"
> > exchanges[exchangeId] = new ccxtpro[exchangeId]({
> > enableRateLimit: true,
> > options: {
> > defaultType: 'option',
> > adjustForTimeDifference: true,
> > keepAlive: true,
> > pingInterval: 5000,
> > pingTimeout: 20000,
> > },
> > });
> 
> Have you found a solution?

Currently, I automatically reconnect within a few seconds after a drop. I also send periodic pong messages, but the connections still get terminated, most likely because of the excessive number of WebSocket connections from a single IP. So far, auto-reconnect is the only workaround that works for me. If you’ve found a better solution, I’d love to know.

**Comment by kroitor** [comment]
Closing this long running thread during a cleanup with the definitive answer, since this is where search engines send everyone who hits the error. The RequestTimeout on a missing ping pong is the detection working, not the defect: a connection that stopped receiving keepalives is surfaced instead of hanging silently on stale data, which is the failure mode people reported before this existed. The supported pattern is to catch RequestTimeout in the watch loop and simply continue, the next watch call resubscribes over a fresh connection. The most common root cause of the timeouts themselves is the event loop being starved by heavy work between watch calls, the consuming loop should call watch again immediately and offload processing elsewhere. The keepalive scheduling and reconnection logic have also been reworked repeatedly since the 2.x era of this thread, including the orderbook resync behavior that the okx staleness reports here touched. If timeouts still occur on a current version with a tight consume loop and an unblocked event loop, a fresh issue with a verbose log would be very welcome. Thanks to everyone who contributed diagnostics here over the years! See also the sibling thread https://github.com/ccxt/ccxt/issues/16123.
